internal/charmstore/debug.go

```diff
@@ -15,8 +18,75 @@ func serveDebugInfo(http.Header, *http.Request) (interface{}, error) {
 	return appver.VersionInfo, nil
 }
 
-func newServiceDebugHandler() http.Handler {
+// GET /debug/check.
+func debugCheck(checks map[string]func() error) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		n := len(checks)
+		type result struct {
+			name string
+			err  error
+		}
+		c := make(chan result)
+		for name, check := range checks {
+			go func(n string, check func() error) {
+				c <- result{name: n, err: check()}
+			}(name, check)
+		}
+		results := make(map[string]string, n)
+		var failed bool
+		for n > 0 {
+			res := <-c
+			if res.err == nil {
+				results[res.name] = "OK"
+			} else {
+				failed = true
+				results[res.name] = res.err.Error()
+			}
+			n--
+		}
+		w.Header().Set("Content-Type", "application/json")
+		if failed {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		enc := json.NewEncoder(w)
+		if err := enc.Encode(results); err != nil {
```

Comment:
Although I like the idea of producing the same body regardless of
error or success, I do feel this somewhat breaks our otherwise universal
contract that if we produce an error, it looks like a params.Error.

That's why I suggested the error message form originally.

Another possibility would be to add a field to the Error return type
for this one call.

   type error struct { 
       params.Error
       Info struct {
            Results map[string] string
       }
    }